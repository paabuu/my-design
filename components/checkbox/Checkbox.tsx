import * as React from 'react';
import * as classNames from 'classnames';

import './style/index.less';

export interface CheckboxProps {
    prefixCls?: string;
    className?: string;
    defaultChecked?: boolean;
    style?: React.CSSProperties;
    disabled?: boolean;
    checked?: boolean;
    onChange?: (e: React.ChangeEvent<HTMLInputElement>) => void;
    onClick?: React.MouseEventHandler<any>;
    onMouseEnter?: React.MouseEventHandler<any>;
    onMouseLeave?: React.MouseEventHandler<any>;
    onKeyPress?: React.KeyboardEventHandler<any>;
    onKeyDown?: React.KeyboardEventHandler<any>;
    value?: any;
    tabIndex?: number;
    name?: string;
    children?: React.ReactNode | string;
}

// export interface CheckboxProps extends AbstractCheckboxProps<CheckboxChangeEvent> {
//     indeterminate?: boolean;
// }

// export interface CheckboxChangeEventTarget extends CheckboxProps {
//     checked: boolean;
// }

// export interface CheckboxChangeEvent {
//     target: CheckboxChangeEventTarget;
//     stopPropagation: () => void;
//     preventDefault: () => void;
//     nativeEvent: MouseEvent;
// }

export interface State {
    checked: boolean;
}

export default class Checkbox extends React.Component<CheckboxProps, State> {
    constructor(props: CheckboxProps) {
        super(props);
        this.state = {
            checked: !!props.checked
        };
    }

    static getDrivedStateFromProps(props: CheckboxProps) {
        if (props.checked !== undefined) {
            return { checked: props.checked };
        }
        
        return null;
    }

    handleOnClick = () => {
        if (this.props.checked === undefined) {
            this.setState({ checked: !this.state.checked });
        } 
    }

    handleOnChange = (e: React.ChangeEvent<HTMLInputElement>) => {
        this.props.onChange && this.props.onChange(e);
    }
    
    render() {
        const { className, onChange, children, ...rest } = this.props;
        const { checked } = this.state;
        const classnames = classNames(className, 'md-checkbox', {
            "md-checkbox-checked": checked,
            "md-checkbox-disabled": rest.disabled
        });
        
        return (
            <label className="md-checkbox-wrapper">
                <span className={classnames}>
                    <input 
                        type="checkbox"
                        checked={checked}
                        onClick={this.handleOnClick}
                        onChange={this.handleOnChange}
                        {...rest}
                    />
                    <span className="md-checkbox-inner"></span>
                </span>
                <span>{this.props.children}</span>
            </label>
        );
    }
}