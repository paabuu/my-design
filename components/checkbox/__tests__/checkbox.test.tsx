import 'jsdom-global/register';
import * as React from 'react';
import { render, mount } from 'enzyme';
import Checkbox from '..';

describe('Checkbox', () => {
    it('render correctly', () => {
        const wrapper1 = render(<Checkbox>Checkbox</Checkbox>);
        expect(wrapper1).toMatchSnapshot();
        const wrapper2 = render(<Checkbox checked={true}>Checked</Checkbox>);
        expect(wrapper2).toMatchSnapshot();
        const wrapper3 = render(<Checkbox disabled={false}>Disabled</Checkbox>);
        expect(wrapper3).toMatchSnapshot();
        const wrapper4 = render(<Checkbox className="new-class">New Class</Checkbox>);
        expect(wrapper4).toMatchSnapshot();
    });

    it('toggle checked status when change', () => {
        const wrapper = mount(<Checkbox>Checkbox</Checkbox>);
        expect(wrapper).toMatchSnapshot();
        wrapper.find('input').simulate('click');
        expect(wrapper).toMatchSnapshot();
    });
});