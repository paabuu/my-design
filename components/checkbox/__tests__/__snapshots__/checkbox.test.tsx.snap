// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox render correctly 1`] = `
<label
  class="md-checkbox-wrapper"
>
  <span
    class="md-checkbox"
  >
    <input
      type="checkbox"
    />
    <span
      class="md-checkbox-inner"
    />
  </span>
  <span>
    Checkbox
  </span>
</label>
`;

exports[`Checkbox render correctly 2`] = `
<label
  class="md-checkbox-wrapper"
>
  <span
    class="md-checkbox md-checkbox-checked"
  >
    <input
      checked=""
      type="checkbox"
    />
    <span
      class="md-checkbox-inner"
    />
  </span>
  <span>
    Checked
  </span>
</label>
`;

exports[`Checkbox render correctly 3`] = `
<label
  class="md-checkbox-wrapper"
>
  <span
    class="md-checkbox"
  >
    <input
      type="checkbox"
    />
    <span
      class="md-checkbox-inner"
    />
  </span>
  <span>
    Disabled
  </span>
</label>
`;

exports[`Checkbox render correctly 4`] = `
<label
  class="md-checkbox-wrapper"
>
  <span
    class="new-class md-checkbox"
  >
    <input
      type="checkbox"
    />
    <span
      class="md-checkbox-inner"
    />
  </span>
  <span>
    New Class
  </span>
</label>
`;

exports[`Checkbox toggle checked status when change 1`] = `
<Checkbox>
  <label
    className="md-checkbox-wrapper"
  >
    <span
      className="md-checkbox"
    >
      <input
        checked={false}
        onChange={[Function]}
        onClick={[Function]}
        type="checkbox"
      />
      <span
        className="md-checkbox-inner"
      />
    </span>
    <span>
      Checkbox
    </span>
  </label>
</Checkbox>
`;

exports[`Checkbox toggle checked status when change 2`] = `
<Checkbox>
  <label
    className="md-checkbox-wrapper"
  >
    <span
      className="md-checkbox md-checkbox-checked"
    >
      <input
        checked={true}
        onChange={[Function]}
        onClick={[Function]}
        type="checkbox"
      />
      <span
        className="md-checkbox-inner"
      />
    </span>
    <span>
      Checkbox
    </span>
  </label>
</Checkbox>
`;
